core/sail/api/src/main/java/org/eclipse/rdf4j/sail/helpers/AbstractSailConnection.java

```diff
@@ -375,61 +385,70 @@ public final void prepare() throws SailException {
 
 	@Override
 	public final void commit() throws SailException {
-		if (isActive()) {
-			endUpdate(null);
-		}
-		connectionLock.readLock().lock();
 		try {
-			verifyIsOpen();
+			if (isActive()) {
+				endUpdate(null);
+			}
 
-			updateLock.lock();
+			connectionLock.readLock().lock();
 			try {
-				if (txnActive) {
-					if (!txnPrepared) {
-						prepareInternal();
+				verifyIsOpen();
+
+				updateLock.lock();
+				try {
+					if (txnActive) {
+						if (!txnPrepared) {
+							prepareInternal();
+						}
+						commitInternal();
+						txnActive = false;
+						txnPrepared = false;
 					}
-					commitInternal();
-					txnActive = false;
-					txnPrepared = false;
+				} finally {
+					updateLock.unlock();
 				}
 			} finally {
-				updateLock.unlock();
+				connectionLock.readLock().unlock();
 			}
 		} finally {
-			connectionLock.readLock().unlock();
+			transactionSettings.clear();
```

Comment:
This seems like it will cause threadsafety issues as the implementation is a non-threadsafe Map and this is outside of the lock. Is there a reason why the transaction settings cannot be cleared from within the lock section?

Reply:
That code is actually from the initial implementation, will remove.